lib/ecto/adapters/sql.ex

```diff
@@ -311,17 +326,50 @@ defmodule Ecto.Adapters.SQL do
   end
 
   @doc false
+  def prepare_execute(repo, _meta, prepared, params, nil, opts) do
+    {query, result} = sql_call!(repo, :prepare_execute, prepared, params, nil, opts)
+    %{rows: rows, num_rows: num} = result
+    {query, {num, rows}}
+  end
+
+  def prepare_execute(repo, %{select: %{fields: fields}}, prepared, params, preprocess, opts) do
+    mapper = &process_row(&1, preprocess, fields)
+    {query, result} = sql_call!(repo, :prepare_execute, prepared, params, mapper, opts)
+    %{rows: rows, num_rows: num} = result
+    {query, {num, rows}}
+  end
+
+  @doc false
+  def execute(repo, _meta, %{} = prepared, params, nil, opts) do
```

Comment:
Here we only match on map, when the query has been prepared and call execute. We are assuming the query is a struct because of DBConnection.Query. We could tag database prepared queries with a tuple to make the code clearer, e.g.`{:prepared, query}`.